freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -102,15 +104,25 @@
     @Inject
     private WorkloadCredentialService workloadCredentialService;
 
+    @Inject
+    private CommonPermissionCheckingUtils commonPermissionCheckingUtils;
+
     public Operation synchronizeUsers(String accountId, String actorCrn, Set<String> environmentCrnFilter,
             Set<String> userCrnFilter, Set<String> machineUserCrnFilter) {
+        List<Stack> stacks = getStacksForSync(accountId, actorCrn, environmentCrnFilter, userCrnFilter, machineUserCrnFilter);
+        return performSyncForStacks(accountId, actorCrn, environmentCrnFilter, userCrnFilter, machineUserCrnFilter, stacks);
+    }
 
-        validateParameters(accountId, actorCrn, environmentCrnFilter, userCrnFilter, machineUserCrnFilter);
-        LOGGER.debug("Synchronizing users in account {} for environmentCrns {}, userCrns {}, and machineUserCrns {}",
-                accountId, environmentCrnFilter, userCrnFilter, machineUserCrnFilter);
+    public Operation synchronizeUsersWithCustomPermissionCheck(String accountId, String actorCrn, Set<String> environmentCrnFilter,
+            Set<String> userCrnFilter, Set<String> machineUserCrnFilter, AuthorizationResourceAction action) {
+        List<Stack> stacks = getStacksForSync(accountId, actorCrn, environmentCrnFilter, userCrnFilter, machineUserCrnFilter);
+        List<String> relatedEnvironmentCrns = stacks.stream().map(stack -> stack.getEnvironmentCrn()).collect(Collectors.toList());
+        commonPermissionCheckingUtils.checkPermissionForUserOnResources(action, actorCrn, relatedEnvironmentCrns);
+        return performSyncForStacks(accountId, actorCrn, environmentCrnFilter, userCrnFilter, machineUserCrnFilter, stacks);
+    }
 
-        List<Stack> stacks = stackService.getMultipleByEnvironmentCrnOrChildEnvironmantCrnAndAccountId(environmentCrnFilter, accountId);
-        LOGGER.debug("Found {} stacks", stacks.size());
+    private Operation performSyncForStacks(String accountId, String actorCrn, Set<String> environmentCrnFilter,
```

Comment:
I don't think we need the environmentCrnFilter here because we've already gotten the relevant stacks.

Reply:
ok!

Reply:
fixed